codalab/worker/download_util.py

```diff
@@ -125,9 +155,100 @@ def _compute_target_info(path, depth):
         result['type'] = 'directory'
         if depth > 0:
             result['contents'] = [
-                _compute_target_info(os.path.join(path, file_name), depth - 1)
+                _compute_target_info_local(os.path.join(path, file_name), depth - 1)
                 for file_name in os.listdir(path)
             ]
     if result is None:
         raise PathException()
     return result
+
+
+def _compute_target_info_beam(path, depth):
+    """Computes target info for a file that is externalized on a location
+    such as Azure, by using the Apache Beam FileSystem APIs."""
+
+    linked_bundle_path = parse_linked_bundle_url(path)
+    if not FileSystems.exists(linked_bundle_path.bundle_path):
+        raise PathException(linked_bundle_path.bundle_path)
+    if not linked_bundle_path.is_archive:
+        # Single file
+        file = FileSystems.match([path])[0].metadata_list[0]
+        return {
+            'name': linked_bundle_path.bundle_uuid,
+            'type': 'file',
+            'size': file.size_in_bytes,
+            'perm': 0o755,
```

Comment:
We're just making up these permissions?  Shouldn't this come from the actual bundle?

Reply:
This code path is actually never reached, because everything on Azure Blob Storage is stored as an archive. I'll raise an exception instead.

Reply:
Done